src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -325,6 +329,91 @@ EnsureCitusTableCanBeCreated(Oid relationOid)
 }
 
 
+/*
+ * undistribute_table gets a distributed table name and
+ * udistributes the table.
+ */
+Datum
+undistribute_table(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+	EnsureCoordinator();
+
+	Oid relationId = PG_GETARG_OID(0);
+	const char *relationName = get_rel_name(relationId);
+
+	EnsureTableOwner(relationId);
+
+	Relation relation = relation_open(relationId, ExclusiveLock);
+
+	if (TableReferencing(relationId))
+	{
+		ereport(ERROR, (errmsg("Cannot undistribute table "
+							   "because it has a foreign key.")));
+	}
+
+	if (TableReferenced(relationId))
+	{
+		ereport(ERROR, (errmsg("Cannot undistribute table "
+							   "because a foreign key references to it.")));
+	}
+
+	StringInfo query = makeStringInfo();
+
+	SPI_connect();
+
+	List *tableDDLEvents = GetTableDDLEvents(relationId, true);
+	char *createTableQuery = linitial(tableDDLEvents);
+	list_delete_first(tableDDLEvents);
+
+	Node *parseTree = ParseTreeNode(createTableQuery);
+
+	((CreateStmt *) parseTree)->relation->relname = "tmp1";
+
+	CitusProcessUtility(parseTree, createTableQuery, PROCESS_UTILITY_TOPLEVEL, NULL,
+						None_Receiver, NULL);
+
+	resetStringInfo(query);
+	appendStringInfo(query, "INSERT INTO tmp1 SELECT * FROM %s", relationName);
+	SPI_execute(query->data, false, 0);
+
+	ObjectAddress undistributeTableAddres;
+
+	undistributeTableAddres.classId = RelationRelationId;
+	undistributeTableAddres.objectId = relationId;
+	undistributeTableAddres.objectSubId = 0;
+
+	DirectFunctionCall3(master_drop_all_shards,
```

Comment:
I'm not sure about it just asking:, Is it safe to use SPI and then use `DirectFunctionCall3 `, and have them all in the same TX?

Reply:
I think it's fine based on this part of the docs: https://www.postgresql.org/docs/11/spi-spi-start-transaction.html 
Most importantly this bit:
> Normally, when an SPI-using procedure is called, there is already a transaction active

This suggests that it's simply running in the same transaction as that the UDF was called with. So as long as you do not commit manually it seems fine.

Reply:
fwiw, we have a ExecuteQueryStringIntoDestReceiver function which you can give a query string and use with None_Receiver.